lib.rs

```diff
@@ -65,17 +116,19 @@ impl<D: Decorator> Format<D> {
         }
     }
 
+    // Returns `true` if message is empty
     fn print_msg_header(&self,
                         io: &mut io::Write,
                         rd: &D::RecordDecorator,
                         record: &Record)
-                        -> io::Result<()> {
+                        -> io::Result<bool> {
         try!(rd.fmt_timestamp(io, &*self.fn_timestamp));
         try!(rd.fmt_level(io,
                           &|io: &mut io::Write| write!(io, " {} ", record.level().as_short_str())));
 
-        try!(rd.fmt_msg(io, &|io| write!(io, "{}", record.msg())));
-        Ok(())
+        let mut writer = CountingWriter::new(io, true);
+        try!(rd.fmt_msg(&mut writer, &|io| write!(io, "{}", record.msg())));
```

Comment:
> Don't you just want to use...

I definitely would! The problem is that `fmt_msg` takes the closure as a `Fn`, not a `FnOnce`, and therefore you can't write an outer variable within the closure without synchronization primitive. An AtomicBool probably wouldn't add too much overhead, and it would get rid of the `csi` code.

> If so, it seems to me that it would be a good idea to not print control sequences at all in fmt_msg if there was no output.

There too, it would be really nice, but that would require going back to the write-twice approach because the `Arguments` type exposes nothing that tells us if the message is empty.